service/raft/raft_group0.cc

```diff
@@ -400,19 +419,62 @@ future<> raft_group0::setup_group0(db::system_keyspace& sys_ks, const std::unord
     group0_log.info("setup_group0: joining group 0...");
     co_await join_group0(std::move(initial_contacts_as_raft_addrs), false /* non-voter */);
     group0_log.info("setup_group0: successfully joined group 0.");
+
+    // Enter `synchronize` upgrade state in case the cluster we're joining has recently enabled Raft
+    // and is currently in the middle of `upgrade_to_group0()`. For that procedure to finish
+    // every member of group 0 (now including us) needs to enter `synchronize` state.
+    co_await _client.set_group0_upgrade_state(group0_upgrade_state::synchronize);
+
+    // Wait until one of group 0 members enters `group0_upgrade_state::use_post_raft_procedures`
+    // or all members enter `group0_upgrade_state::synchronize`.
+    //
+    // In a fully upgraded cluster this should finish immediately (if the network works well) - everyone is in `use_post_raft_procedures`.
+    // In a cluster that is currently in the middle of `upgrade_to_group0`, this will cause us to wait until the procedure finishes.
+    group0_log.info("setup_group0: ensuring that the cluster has fully upgraded to use Raft...");
+    co_await wait_for_peers_to_enter_synchronize_state(_raft_gr.group0(), _ms, _abort_source, _shutdown_gate.hold());
+
+    group0_log.info("setup_group0: the cluster is ready to use Raft. Finishing.");
+    co_await _client.set_group0_upgrade_state(group0_upgrade_state::use_post_raft_procedures);
```

Comment:
We're jumping from `synchronize` state. If a remote entered `use_post_raft_procedures`, they must have verified that we were in `synchronize` and that schema was syncd on all nodes after we were in `synchronize`. So the schema must have been synchronized - we just depend on someone else to verify that for us.

Reply:
Ugh, now I noticed you're commenting on the bootstrap code.

Indeed this doesn't make sense.

It should follow the regular upgrade path - enter synchronize, and either
- finish if someone entered `use_post_raft_procedures`
- or wait until everyone entered `synchronize` and then wait until schema is synchronized.

Just because everyone entered `synchronize` doesn't mean the upgrade has finished.

The design doc has this bug as well:
```
- Set upgrade_state = synchronize.
- Wait until every member of group 0 enters synchronize or some member enters use_new_procedures.
- Set upgrade_state = use_new_procedures.
```
It completely flew over my head.

Reply:
I fixed the doc, I'll fix this code tomorrow.

Reply:
Fixed